### ISSUE TITLE
petite & proxetta : problem with mixing scopes (singleton & prototype)

### ISSUE THREAD

**Issue by moh-sushi** [oblac/jodd#343]
Hi,

i am having an issue while mixing the scopes prototype and singleton.

my situation:
the petite continer should wire petite beans into a non petite bean instance.

non petite beans
||
||
\/
main petite bean (singleton)
||
||
\/
sub petite bean (prototype)

Follwoing exception is thrown:

```
Caused by: jodd.proxetta.ProxettaException: Target missing
    at jodd.proxetta.ProxettaBuilder.process(ProxettaBuilder.java:192)
    at jodd.proxetta.ProxettaBuilder.define(ProxettaBuilder.java:247)
    at jodd.petite.ScopedProxyManager.createScopedProxyBean(ScopedProxyManager.java:145)
    at jodd.petite.ScopedProxyManager.lookupValue(ScopedProxyManager.java:105)
    at jodd.petite.PetiteContainer.wireProperties(PetiteContainer.java:157)
    at jodd.petite.PetiteContainer.wireBean(PetiteContainer.java:134)
    at jodd.petite.PetiteContainer.registerBeanAndWireAndInjectParamsAndInvokeInitMethods(PetiteContainer.java:380)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:366)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:328)
    at jodd.petite.PetiteContainer.wireProperties(PetiteContainer.java:162)
    at jodd.petite.PetiteContainer.wireBean(PetiteContainer.java:134)
    at jodd.petite.PetiteContainer.registerBeanAndWireAndInjectParamsAndInvokeInitMethods(PetiteContainer.java:380)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:366)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:328)
    at jodd.petite.PetiteContainer.wireProperties(PetiteContainer.java:162)
    at jodd.petite.PetiteContainer.wireBean(PetiteContainer.java:134)
    at jodd.petite.PetiteContainer.registerBeanAndWireAndInjectParamsAndInvokeInitMethods(PetiteContainer.java:380)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:366)
    at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:317)
```

I have attached an example with unit test.
[jodd petite proxetta - mixing scopes.zip](https://github.com/oblac/jodd/files/528966/jodd.petite.proxetta.-.mixing.scopes.zip)

Is this a bug or usage failure on my side?
I am using jodd 3.7.1

Thank you!

bye
Sascha

**Comment by moh-sushi** [comment]
`	ProxettaBuilder 
	
	with 'class jodd.issue.service.xxx.impl.OtherPetiteBean$$Proxetta'
	/**
	 * Defines class as a target.
	 */
	protected void setTarget(Class target) {
		checkTarget();

		try {
			targetInputStream = ClassLoaderUtil.getClassAsStream(target); // <-- null (see below)
			targetClass = target;
			targetClassName = target.getName();
		} catch (IOException ioex) {
			StreamUtil.close(targetInputStream);
			throw new ProxettaException("Unable to stream class: " + target.getName(), ioex);
		}
	}



	/**
	 * Opens a resource of the specified name for reading.
	 * @see #getResourceUrl(String, ClassLoader)
	 */
	public static InputStream getResourceAsStream(String resourceName, ClassLoader callingClass) throws IOException {
		URL url = getResourceUrl(resourceName, callingClass); // <!-- URL will be null!
		if (url != null) {
			return url.openStream();
		}
		return null;
	}`

My very first investigation - maybe useful?! :-)

**Comment by igorspasic** [comment]
Hello! This seems like target class is not found.

**Comment by igorspasic** [comment]
OK, here is what is going on. `Sub` class is first 'proxified' on creation, and then we want to proxify again, for the mixed-scope purpose. So we want to proxify a class on two places; second time it fails.

Let me see if I can fix this by detecting this case and only proxifying once, with two aspects.

**Comment by moh-sushi** [comment]
Hi,

i am still getting a failure:

`
jodd.proxetta.ProxettaException: Target class not found:   de.mohsushi.jodd.example1.impl.SubPetiteBean$$Proxetta
	at jodd.proxetta.ProxettaBuilder.setTarget(ProxettaBuilder.java:132)
	at jodd.proxetta.impl.ProxyProxettaBuilder.setTarget(ProxyProxettaBuilder.java:60)
	at jodd.proxetta.impl.ProxyProxetta.builder(ProxyProxetta.java:72)
	at jodd.petite.ScopedProxyManager.createScopedProxyBean(ScopedProxyManager.java:163)
	at jodd.petite.ScopedProxyManager.lookupValue(ScopedProxyManager.java:107)
	at jodd.petite.PetiteContainer.wireProperties(PetiteContainer.java:157)
	at jodd.petite.PetiteContainer.wireBean(PetiteContainer.java:134)
	at jodd.petite.PetiteContainer.registerBeanAndWireAndInjectParamsAndInvokeInitMethods(PetiteContainer.java:380)
	at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:366)
	at jodd.petite.PetiteContainer.getBean(PetiteContainer.java:328)
	at jodd.petite.PetiteContainer.wireProperties(PetiteContainer.java:162)
	at jodd.petite.PetiteContainer.wireBean(PetiteContainer.java:134)
	at jodd.petite.PetiteContainer.registerBeanAndWireAndInjectParamsAndInvokeInitMethods(PetiteContainer.java:380)
	at jodd.petite.PetiteContainer.wire(PetiteContainer.java:403)
	at jodd.petite.PetiteContainer.wire(PetiteContainer.java:393)
	at de.mohsushi.jodd.example1.TestForExample1.testWithMixingScopesSingletonAndProto(TestForExample1.java:40)
`

Tasks to reproduce :
- download attached [zip-file](https://github.com/oblac/jodd/files/528966/jodd.petite.proxetta.-.mixing.scopes.zip) (1st post)
- unzip
- change java.version in pom.xml  -> 1.8
- change jodd.version in pom.xml -> 3.8.0
- execute test

Can you verify my problem?

**Comment by igorspasic** [comment]
Ok, let me try again, I will reopen the issue.

**Comment by igorspasic** [comment]
Just checked. Please change line **24** in `TestForExample1` to:

     petiteContainer = new ProxettaAwarePetiteContainer(proxyProxetta, petiteConfig);

as this one has been changed and it does what it needs to do :)

Thanx!

**Comment by moh-sushi** [comment]
Hi,

thank you for your hint.
Done.

But ....  

There is another issue with "mixing singleton and proto scope" therefore i comment again :-)

My current situation :

non petite bean
||
||
/
main petite bean (singleton)
||
||
/
sub petite bean delegate (prototype) (delegates all methods to external bean)
||
||
/
external bean (an external bean with no jodd dependency; stateful bean; login/logout)

> Prototype scope doesn't pool any beans, so each time bean is requested, a new instance will be created. Prototype scope does not call destroy methods.

my understanding:
The ProtoScope-Bean is only once requested - while wiring the MainPetiteBean.
Each method call of this ProtoScope-Bean should remain to this injected instance.
But after each method call another instance of the ProtoScope bean is available.

Does your ProtoScope work as expected that is my usage (and my understanding) of ProtoScope within MainPetiteBean is wrong?

see updated test case / test project -> 
[jodd petite proxetta - mixing scopes.zip](https://github.com/oblac/jodd/files/566770/jodd.petite.proxetta.-.mixing.scopes.zip)

Thank you!

**Comment by moh-sushi** [comment]
First, thank you for your answer.

I wil try to change behaviour by using the linked `PetiteProvider`.

regarding your question:
Well, good question....
> I might check how other IOCs works :)))

I will try to setup testcase with [Spring](https://spring.io/) today.

**Comment by moh-sushi** [comment]
First, thank you for your answer.

I will try to change it by using `PetiteProvider`.

> I might check how other IOCs works :)))

I try to setup testcase with [Spring](http://spring.io/) today.

**Comment by igorspasic** [comment]
Hey @moh-sushi! This is a question on how the mixed scopes works. `Protoscope` is maybe one scope that I am not quite sure how it should behave. Let me explain your example in the world of mixed scopes.

Let's say we have 2 beans: `Main` (singleton) and  `Sub`:

    Main <--- Sub

For this example, lets say that `Sub` is **session** scope (same http session will have same instance). Injecting session into singleton is mixed scope. Every time when `Main` singleton object calls a method from `sub`, we need to actually get the `sub` bean instance from the session scope and call method on it. This is what the proxy is used for.

Now let's say that `Sub` bean is a protoscope. If you follow above logic, it would mean that _everytime_ when you operate on a `sub`, an instance would be looked up - and since this is a protoscope, instances are not pooled, so: _new instance will be created every time_. You will get this:

    sub.hello()

would be executed effectively as:

   new Sub().hello()

This explains current behavior. But I must admin that when you said:

> The ProtoScope-Bean is only once requested - while wiring the MainPetiteBean.

it also feels correct! 👌 

Lets see what we can do:

+ **/A** I can change the behavior of protoscope in mixed environment, to match what you said.

+ **/B** you might use `PetiteProvider` ([doc](http://jodd.org/doc/petite/petite-provider.html)).

@moh-sushi what do you think behavior for case /A should be? I might check how other IOCs works :)))

**Comment by moh-sushi** [comment]
Hi,

update testcase -> usage of spring framework. 
[jodd petite proxetta - mixing scopes.zip](https://github.com/oblac/jodd/files/574009/jodd.petite.proxetta.-.mixing.scopes.zip)

Please, check my spring test config.
It should work properly.

You can see that the requested prototype instance remains the same one.
There is a further doc regarding different scopes -> http://javabeat.net/spring-bean-scopes-singleton-prototype/


Did you read other ioc documentaion? 
What do you think?

**Comment by igorspasic** [comment]
I was reading the other ioc docs, but I remember I was focused on http session scope :) Okey, so this means the following:

+ It is ok to inject PROTO scope to _any_ other scope (there is no mixed scope)
+ However, for every injection the new instance of PROTO bean will be created (that is different from SINGLETON scope, where the same instance is shared instead)

wdyt @moh-sushi?

The change is very simple, and I will push it today, just ping me back to see if we are aligned :) Here is, for example, the output of your example - no exceptions:

```
    before login
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after login
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    before execute
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after execute
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    before logout
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after logout
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539
Disconnected from the target VM, address: '127.0.0.1:53253', transport: 'socket'

executing non jodd petite bean -> de.mohsushi.jodd.example1.ExternalBean
 -- -- -- next call of external bean

    before login
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after login
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    before execute
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after execute
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    before logout
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539


    after logout
de.mohsushi.jodd.example1.impl.MainPetiteBean$$Proxetta | 1254157669
de.mohsushi.jodd.example1.impl.SubPetiteBeanDelegate$$Proxetta | 1551660539

executing non jodd petite bean -> de.mohsushi.jodd.example1.ExternalBean
```

Does this look ok?

**Comment by igorspasic** [comment]
Hey @moh-sushi check it now :)

**Comment by moh-sushi** [comment]
> wdyt @moh-sushi?

sounds ok.

> Does this look ok?

yes - looks ok.
